Improve performance of IVsLanguageDebugInfo.GetNameOfLocation...

KevinH-MS opened the pull request:
**Bug number** - Fixes #1977
**Customer scenario** - In large solutions, you may see the following dialog appear for long periods of time when opening your solution or setting new breakpoints (sometimes in excess of a minute):

![determiningname](https://cloud.githubusercontent.com/assets/6464209/7849750/b1608234-048d-11e5-9682-f30d8e2280bb.png)

**Fix description** - Previously, we would fetch a SemanticModel and build containing Symbols for each breakpoint location.  This could be very expensive in a large solution.  However, for the purposes of providing parity with the old implementation, it is adequate to just inspect Syntax.
**Testing done** - Added more comprehensive tests (56 new), ad-hoc, basic perf testing...  I'm not able to consistently repro the "worse case", but even in the best case, the new implementation is twice as fast (20ms vs 43ms when opening Roslyn.sln with 2 breakpoints set).

KevinH-MS commented:
FYI @ManishJayaswal @cston @amcasey ... sending C# implementation out for review while I work on VB...

cston commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -8,33 +10,16 @@
-                miscellaneousOptions:
-                    SymbolDisplayMiscellaneousOptions.EscapeKeywordIdentifiers |
-                    SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
+        private static ObjectPool<StringBuilder> s_stringBuilderPool = SharedPools.Default<System.Text.StringBuilder>();
```
static readonly?

KevinH-MS replied:
Yes

KevinH-MS commented:
FYI @Pilchie for Organizing change (feel free to tag any other appropriate reviewers)...

amcasey commented on src/Features/CSharp/Organizing/Organizers/MemberDeclarationsOrganizer.Comparer.cs:
```diff
@@ -99,8 +99,8 @@ public int Compare(MemberDeclarationSyntax x, MemberDeclarationSyntax y)
+                var xName = ShouldCompareByName(x) ? x.GetNameToken() : default(SyntaxToken);
+                var yName = ShouldCompareByName(y) ? y.GetNameToken() : default(SyntaxToken);
 
                 if ((compare = TokenComparer.NormalInstance.Compare(xName, yName)) != 0)
```
Does this put unnamed members before or after other members?

KevinH-MS replied:
I'm not sure--my only intent is that the new code does what the old code did (after changes to ```GetNameToken```).

amcasey commented on src/Features/CSharp/Organizing/Organizers/MemberDeclarationsOrganizer.Comparer.cs:
```diff
@@ -171,6 +171,21 @@ private OuterOrdering GetOuterOrdering(MemberDeclarationSyntax x)
+                    case SyntaxKind.ConstructorDeclaration:
+                    case SyntaxKind.DestructorDeclaration:
+                    case SyntaxKind.IndexerDeclaration:
+                    case SyntaxKind.OperatorDeclaration:
```
Does this cover conversions?

KevinH-MS replied:
I could add it explicitly here, but it's not necessary, because GetNameToken will return default(SyntaxToken) for it.

amcasey replied:
Maybe add a comment to that effect?

Pilchie replied:
This code is currently dead, so I'm not terribly worried about it.

KevinH-MS replied:
Updated comment.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -8,33 +10,16 @@
-                miscellaneousOptions:
-                    SymbolDisplayMiscellaneousOptions.EscapeKeywordIdentifiers |
-                    SymbolDisplayMiscellaneousOptions.UseSpecialTypes);
+        private static readonly ObjectPool<StringBuilder> s_stringBuilderPool = SharedPools.Default<StringBuilder>();
```
Is PooledStringBuilder unavailable here?

KevinH-MS replied:
Huh...good catch.  I assumed it was not (because its a compiler internal, and I thought we were only allowed to call public APIs in the compiler layer from VS), but it looks like it's linked in as source.  I'll use that instead.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -8,33 +10,16 @@
-
             // Note that we get the current partial solution here.  Technically, this means that we may
             // not fully understand the signature of the member.  But that's ok.  We just need this
             // symbol so we can create a display string to put into the debugger.  If we try to
```
Is this comment still accurate?

KevinH-MS replied:
Upon closer inspection, it would appear that it's not even accurate for the previous code.  It's describing exactly the problem that was manifesting as if it should not occur.  I'll fix it.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
+                else if (nameToken == default(SyntaxToken))
+                {
+                    Debug.Assert(memberDeclaration.Kind() == SyntaxKind.ConversionOperatorDeclaration);
+                    builder.Append((memberDeclaration as ConversionOperatorDeclarationSyntax)?.Type);
```
Is this going to append null if ```memberDeclaration``` isn't a ```ConversionOperatorDeclarationSyntax```?  Should it append ```missingInformationPlaceholder```?

KevinH-MS replied:
Yes, it will result in ```builder.Append(null)``` (which I believe is a no-op).

If that feels too subtle, I can make the null check explicit here.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
+                { 
+                    if (memberDeclaration.Kind() == SyntaxKind.DestructorDeclaration)
+                    {
+                        builder.Append('~');
```
How come this one didn't get a constant?

amcasey replied:
Also, I vaguely recall Paul swapping all such char literals with string literals for perf reasons.  You might want to ask him.

KevinH-MS replied:
I didn't extract this one because it's only used in one place.  I'll ask Paul about the char vs string.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
+                    builder.Append(nameToken.Text);
+                }
+                // parameter list (if any)
+                builder.Append(memberDeclaration.GetParameterList());
```
Would it make sense to pass in the builder?

KevinH-MS replied:
It would be a reasonable optimization, but I'm not sure it's worth it to muck with the other callers of ```GetParameterList```.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
+                // parameter list (if any)
+                builder.Append(memberDeclaration.GetParameterList());
+                // containing type(s)
+                var containingDeclaration = memberDeclaration.GetAncestor<TypeDeclarationSyntax>();
```
Personally, I think this would be clearer with "type" in the name.

KevinH-MS replied:
Done.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
+                    containingDeclaration = containingDeclaration.GetAncestor<TypeDeclarationSyntax>();
+                }
+                // containing namespace(s)
+                if (lastContainer != null)
```
When is this null?

KevinH-MS replied:
Good point--it's not needed in the current code.  Removed...

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
+                        builder.Insert(0, dotTokenText);
+                        var nameSyntax = containingNamespace.Name;
+                        builder.Insert(0, nameSyntax.IsMissing ? missingInformationPlaceholder : nameSyntax.ToString());
+                        containingNamespace = containingNamespace.GetAncestor<NamespaceDeclarationSyntax>();
```
Can it be anything other than the parent?

KevinH-MS replied:
Probably not, but I wasn't sure about broken code.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var lineNumber = text.Lines.GetLineFromPosition(position).LineNumber;
+                var accessor = token.GetAncestor<AccessorDeclarationSyntax>();
+                var memberLine = accessor == null
```
Could put the condition in the argument to ```GetLineFromPosition``` (and use ```?.```).

KevinH-MS replied:
Yes--done.

amcasey commented:
Personally, when I make a change to improve perf, I add a ```// PERF``` comment so no one cleans it up back to the way it used to be.

amcasey commented:
It's not superficially obvious that the new tests give you good code coverage of all those loops and null checks.  Are there are a lot of existing tests?

cston commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
             // If the position does not point to one, try using the first one.
-            if (memberDecl != null &&
-                (memberDecl.Kind() == SyntaxKind.FieldDeclaration || memberDecl.Kind() == SyntaxKind.EventFieldDeclaration))
+            if (memberDeclaration != null &&
```
Perhaps if (memberDeclaration == null) return default(DebugLocationInfo);

KevinH-MS replied:
Good point--changed

cston commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
+                SyntaxNode lastContainer = memberDeclaration;
+                while (containingDeclaration != null)
+                {
+                    builder.Insert(0, dotTokenText);
```
Consider appending the containing types up front rather than inserting them.

cston replied:
Consider appending the containing types up front rather than inserting them. Same for namespaces.

KevinH-MS replied:
I've moved types/namespaces up front, but I'm still inserting them (to avoid having to collect them as I walk up through the ancestors or reverse the list).

cston replied:
Inserting n types and namespaces is O(n^2).

KevinH-MS replied:
Huh...I'd never actually looked at the implementation of StringBuilder until now.  I'd assumed it was some sort of linked list, but nope...backed by array of chars...  I'll change it.

cston commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -43,49 +28,93 @@ internal static class LocationInfoGetter
-                        ? text.Lines.GetLineFromPosition(memberDecl.SpanStart).LineNumber
-                        : text.Lines.GetLineFromPosition(accessor.SpanStart).LineNumber;
+                var text = await document.GetTextAsync(cancellationToken).ConfigureAwait(false);
+                var lineNumber = text.Lines.GetLineFromPosition(position).LineNumber;
```
Consider extracting the name building to a separate non-async method.

KevinH-MS replied:
Done.

KevinH-MS commented:
@amcasey regarding the existing tests--no, the existing tests were not comprehensive (in fact, they were rather sparse).  I do believe that the tests I added provide simple block coverage for the code I added though.

amcasey commented:
:+1:

Pilchie commented:
Hmm - I might extract this code for #2632, since there too I'd like a syntax only representation of a member name.

KevinH-MS commented:
@cston @amcasey VB is done now (if you could take a look at the second commit in the PR)

cston commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -1,94 +1,155 @@
+        {
+            const string missingInformationPlaceholder = "?";
+
+            ArrayBuilder<string> containingDeclarationNames = null;
```
Why not allocate this here? Won't there be a containing declaration in all cases?

KevinH-MS replied:
Top level code (scripting) and broken code...it may or may not be worth it to optimize for those cases.

KevinH-MS replied:
In VB, this is also hit if you have a breakpoint on a field initializer in a class with no containing namespace.

cston replied:
Agreed, it's not worth complicating the code for those cases and it's not clear there'd be a performance benefit.

KevinH-MS replied:
Okay--I'll move the allocation to the caller.

cston commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -1,94 +1,155 @@
+            ref TDeclarationSyntax declaration,
+            Func<TDeclarationSyntax, string> getText,
+            Func<TDeclarationSyntax, TDeclarationSyntax> getAncestor) where TDeclarationSyntax : MemberDeclarationSyntax
+        {
```
Since the method is basically a loop, consider inlining in the callers.

amcasey commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -1,94 +1,155 @@
+            return pooled.ToStringAndFree();
+        }
+
+        private static void AppendContainingDeclarationNames<TDeclarationSyntax>(
```
As we discussed offline, I think this could be more clearly expressed as a single while loop walking up the syntax tree.  You can use an ```ArrayBuilder``` as a stack to avoid allocations.

amcasey commented:
:+1:

amcasey commented:
Thanks!  That's much clearer.

cston commented on src/VisualStudio/VisualBasic/Impl/Debugging/LocationInfoGetter.vb:
```diff
@@ -3,58 +3,117 @@
+                ' containing namespace(s) and type(s)
+                Dim containingDeclarationNames As ArrayBuilder(Of String) = ArrayBuilder(Of String).GetInstance()
+                Dim containingDeclaration = memberDeclaration.Parent
+                If TypeOf memberDeclaration Is TypeStatementSyntax Then
```
Won't the While loop handle this case?

KevinH-MS replied:
It's a special case in VB if you manage to get a breakpoint set at a span that is in a class (but not in one of the members of the class).  In that case, we want to skip the class in the loop below (and just append the name of the class where we do ```' simple name```.

KevinH-MS replied:
Sorry, to further clarify, when this happens, ```memberDeclaration``` will be a ```DeclarationStatementSyntax``` that is parented by the ```TypeBlockSyntax```.  In C#, this doesn't happen, because you don't have a separate block parenting the decl statement.

cston commented on src/VisualStudio/CSharp/Impl/Debugging/LocationInfoGetter.cs:
```diff
@@ -1,94 +1,129 @@
+                    if (type != null)
+                    {
+                        var token = type.GetNameToken();
+                        containingDeclarationNames.Add(token.IsMissing ? missingInformationPlaceholder : token.Text);
```
Should this include type parameters?

KevinH-MS replied:
The old implementation didn't, so I've omitted it here as well.

cston commented:
LGTM

MattGertz commented:
Approved 5/29 by ML Shiproom.
